service/matching/db.go

```diff
@@ -329,7 +339,9 @@ func (db *taskQueueDB) getUserDataLocked() (*persistencespb.VersionedTaskQueueUs
 	case userDataEnabled:
 		return db.userData, db.userDataChanged, nil
 	case userDataDisabled:
-		return nil, nil, errUserDataDisabled
+		// return userDataChanged even with an error here so that a blocking wait can be
+		// interrupted when user data is enabled again.
+		return nil, db.userDataChanged, errUserDataDisabled
```

Comment:
Hmm, maybe. The idea is that this is kind of like a subscription: you call this, you get the current value plus a channel that tells you when the value might have changed. And in the error case, you get a channel that tells you when another call might not return an error anymore. (I suppose to be complete it should close the channel in setUserDataState also.. I'll do that.)

I don't want to expose a separate call to get the changed channel, because it should be atomic with getting the `userData` field (otherwise you could miss an update).

It is a little confusing but not the most confusing thing about this feature...

Also, we can probably remove this whole user-data-disabled mechanism in a couple releases

Reply:
Should we track removing the flag in a GH issue?

Reply:
I'll track it in jira